src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs

```diff
@@ -0,0 +1,158 @@
+        {
+            PackagesLockFile lockFile;
+
+            if (lockFileExists)
+            {
+                lockFile = PackagesLockFileFormat.Read(lockFileName);
+            }
+            else
+            {
+                lockFile = new PackagesLockFile();
+                lockFile.Targets.Add(new PackagesLockFileTarget
+                {
+                    TargetFramework = new NuGetFramework(FrameworkConstants.SpecialIdentifiers.Any)
+                });
+            }
+
+            return lockFile;
+        }
+
+        internal static async Task ApplyChangesAsync(
+            PackagesLockFile lockFile,
+            List<NuGetProjectAction> actionsList,
+            IContentHashUtility contentHashUtil,
+            CancellationToken token)
+        {
+            RemoveUninstalledPackages(lockFile,
+                actionsList.Where(a => a.NuGetProjectActionType == NuGetProjectActionType.Uninstall));
+            await AddInstalledPackagesAsync(lockFile,
+                actionsList.Where(a => a.NuGetProjectActionType == NuGetProjectActionType.Install),
+                contentHashUtil,
+                token);
+        }
+
+        private static void RemoveUninstalledPackages(PackagesLockFile lockFile, IEnumerable<NuGetProjectAction> actionsList)
+        {
+            foreach (var toUninstall in actionsList)
+            {
+                Debug.Assert(toUninstall.NuGetProjectActionType == NuGetProjectActionType.Uninstall);
+
+                foreach (var installed in lockFile.Targets[0].Dependencies)
+                {
+                    if (installed.Id == toUninstall.PackageIdentity.Id)
+                    {
+                        lockFile.Targets[0].Dependencies.Remove(installed);
+                        return;
+                    }
+                }
+            }
+        }
+
+        private static async Task AddInstalledPackagesAsync(
+            PackagesLockFile lockFile,
+            IEnumerable<NuGetProjectAction> actionsList,
+            IContentHashUtility contentHashUtil,
+            CancellationToken token)
+        {
+            foreach (var toInstall in actionsList)
+            {
+                Debug.Assert(toInstall.NuGetProjectActionType == NuGetProjectActionType.Install);
+
+                var newDependency = new LockFileDependency
+                {
+                    Id = toInstall.PackageIdentity.Id,
+                    ContentHash = await contentHashUtil.GetContentHashAsync(toInstall.PackageIdentity, token),
+                    RequestedVersion = new VersionRange(toInstall.PackageIdentity.Version, true, toInstall.PackageIdentity.Version, true),
+                    ResolvedVersion = toInstall.PackageIdentity.Version,
+                    Type = PackageDependencyType.Direct
+                };
+
+                // should keep sorted, but lockFile[Targets[0].Dependencies is an IList<T>, but only List<T> has the .Sort() method, so we have to insert in sorted order.
```

Comment:
why do you need to keep them in sorted order? current impl will really make it hard to see the file diff since it will add new packages at random places which might mess with uninstalling packages.

So I'd suggest to keep it simple to keep appending at the end unless there is a valid reason to do otherwise

Reply:
The PC file is sorted, so it keeps the lock file consistent. And the reason I think it's a good idea is exactly to simplify file diffs. Since the file is JSON, adding a new item at the end will cause 2 lines to diff, the new line and adding a comment to the end of the last item. Also, if a user removes a package, but then installs it again, I don't want their lock file to be different. Keeping it sorted makes sure that the lock file only has a source control diff when the contents are the same, no matter which order they were installed.